File: src/main/java/myratpackexamples/promises/ValidationUtil.java
package myratpackexamples.promises;

import io.vavr.control.Validation;
import ratpack.exec.Promise;

import java.util.function.Function;

import static io.vavr.control.Validation.invalid;
import static ratpack.exec.Promise.value;

public class ValidationUtil {
    public static <E, I, O> Promise<Validation<E, O>> flatMapPromise(
            Validation<E, I> validation, Function<I, Promise<Validation<E, O>>> function) {

        if (validation.isInvalid()) {
            return value(invalid(validation.getError()));
        } else {
            return function.apply(validation.get());
        }
    }
}

File: src/main/java/myratpackexamples/promises/ValidationUtil.kt
package myratpackexamples.promises

import io.vavr.control.Validation
import ratpack.exec.Promise

import java.util.function.Function

import io.vavr.control.Validation.invalid
import ratpack.exec.Promise.value

object ValidationUtil {
    fun <E, I, O> flatMapPromise(
            validation: Validation<E, I>, function: (I) -> Promise<Validation<E, O>>): Promise<Validation<E, O>> {

        return if (validation.isInvalid) {
            value(invalid(validation.error))
        } else {
            function.invoke(validation.get())
        }
    }
}
